File: js/src/denotation.mjs
class Denotation {
  #beginPos;
  #endPos;
  #obj;
  #id;

  constructor(beginPos, endPos, obj, id = null) {
    this.#beginPos = beginPos;
    this.#endPos = endPos;
    this.#obj = obj;
    this.#id = id;
  }

  get beginPos() {
    return this.#beginPos;
  }

  get endPos() {
    return this.#endPos;
  }

  get obj() {
    return this.#obj;
  }

  get id() {
    return this.#id;
  }

  get span() {
    return { begin: this.#beginPos, end: this.#endPos };
  }

  toObject() {
    const result = { span: this.span, obj: this.#obj };
    if (this.#id) {
      result.id = this.#id;
    }
    return result;
  }

  isNestedWithin(other) {
    return other.beginPos <= this.#beginPos && this.#endPos <= other.endPos;
  }

  get isPositionNotInteger() {
    return !(Number.isInteger(this.#beginPos) && Number.isInteger(this.endPos));
  }

  get isPositionNegative() {
    return this.#beginPos < 0 || this.#endPos < 0;
  }

  get isPositionInvalid() {
    return this.#beginPos > this.#endPos;
  }

  isOutOfBounds(textLength) {
    return this.#beginPos >= textLength || this.#endPos > textLength;
  }

  isBoundaryCrossing(other) {
    const startsInsideOther =
      this.#beginPos > other.beginPos && this.#beginPos < other.endPos;
    const endsInsideOther =
      this.#endPos > other.beginPos && this.#endPos < other.endPos;

    return startsInsideOther || endsInsideOther;
  }
}

export default Denotation;

File: ruby/lib/simple_inline_text_annotation/denotation.rb
# frozen_string_literal: true

require_relative "denotation_validator"
require_relative "generator_error"

class SimpleInlineTextAnnotation
  class Denotation
    attr_reader :begin_pos, :end_pos, :obj, :id

    def initialize(begin_pos, end_pos, obj, id = nil)
      @begin_pos = begin_pos
      @end_pos = end_pos
      @obj = obj
      @id = id
    end

    def span
      { begin: @begin_pos, end: @end_pos }
    end

    def to_h
      { id: @id, span: span, obj: @obj }.compact
    end

    def nested_within?(other)
      other.begin_pos <= @begin_pos && @end_pos <= other.end_pos
    end

    def position_not_integer?
      !(@begin_pos.is_a?(Integer) && @end_pos.is_a?(Integer))
    end

    def position_negative?
      @begin_pos.negative? || @end_pos.negative?
    end

    def position_invalid?
      @begin_pos > @end_pos
    end

    def out_of_bounds?(text_length)
      @begin_pos >= text_length || @end_pos > text_length
    end

    def boundary_crossing?(other)
      starts_inside_other = @begin_pos > other.begin_pos && @begin_pos < other.end_pos
      ends_inside_other = @end_pos > other.begin_pos && @end_pos < other.end_pos

      starts_inside_other || ends_inside_other
    end
  end
end
